// @benmalka/foxdriver/build/domains/preference.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _actor = require("../actor");

var _actor2 = _interopRequireDefault(_actor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Actor to manage browser preferences.
 */
class Preference extends _actor2.default {
  /**
   * Get boolean preference.
   *
   * @param  {String}  value    name of preference
   * @return {Promise.Boolean}  preference value
   */
  async getBoolPref(value) {
    const response = await this.request('getBoolPref', {
      value
    });
    return response.value;
  }
  /**
   * Get char preference.
   *
   * @param  {String}  value  name of preference
   * @return {Promise.Char}   preference value
   */


  async getCharPref(value) {
    const response = await this.request('getCharPref', {
      value
    });
    return response.value;
  }
  /**
   * Get integer preference.
   *
   * @param  {String}  value    name of preference
   * @return {Promise.Number}   preference value
   */


  async getIntPref(value) {
    const response = await this.request('getIntPref', {
      value
    });
    return response.value;
  }
  /**
   * Get all preferences.
   *
   * @return {Promise.Object[]}  list of preferences
   */


  async getAllPrefs() {
    const {
      value
    } = await this.request('getAllPrefs');
    return value;
  }
  /**
   * Set boolean preference.
   *
   * @param {String}  name   preference name
   * @param {Boolean} value  preference value
   * @return {Promise}       request response
   */


  setBoolPref(name, value) {
    return this.request('setBoolPref', {
      name,
      value
    });
  }
  /**
   * Set char preference.
   *
   * @param {String}  name   preference name
   * @param {Char}    value  preference value
   * @return {Promise}       request response
   */


  setCharPref(name, value) {
    return this.request('setCharPref', {
      name,
      value
    });
  }
  /**
   * Set integer preference.
   *
   * @param {String}  name   preference name
   * @param {Number}  value  preference value
   * @return {Promise}       request response
   */


  setIntPref(name, value) {
    return this.request('setIntPref', {
      name,
      value
    });
  }
  /**
   * Clear user set preference
   * @param  {Srting} name  name of preference
   * @return {Promise}      request response
   */


  clearUserPref(name) {
    return this.request('clearUserPref', {
      name
    });
  }

}

exports.default = Preference;
module.exports = exports["default"];
