// @packages/rewriter/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.terminateAllWorkers = exports.createInitialWorkers = exports.DeferredSourceMapCache = exports.rewriteHtmlJsAsync = exports.rewriteJsAsync = exports.HtmlJsRewriter = void 0;
var html_1 = require("./html");
Object.defineProperty(exports, "HtmlJsRewriter", { enumerable: true, get: function () { return html_1.HtmlJsRewriter; } });
var async_rewriters_1 = require("./async-rewriters");
Object.defineProperty(exports, "rewriteJsAsync", { enumerable: true, get: function () { return async_rewriters_1.rewriteJsAsync; } });
Object.defineProperty(exports, "rewriteHtmlJsAsync", { enumerable: true, get: function () { return async_rewriters_1.rewriteHtmlJsAsync; } });
var deferred_source_map_cache_1 = require("./deferred-source-map-cache");
Object.defineProperty(exports, "DeferredSourceMapCache", { enumerable: true, get: function () { return deferred_source_map_cache_1.DeferredSourceMapCache; } });
var threads_1 = require("./threads");
Object.defineProperty(exports, "createInitialWorkers", { enumerable: true, get: function () { return threads_1.createInitialWorkers; } });
Object.defineProperty(exports, "terminateAllWorkers", { enumerable: true, get: function () { return threads_1.terminateAllWorkers; } });
